### ISSUE TITLE
[clang][analyzer] False positve in core.NullDereference checker

### ISSUE THREAD

**Issue by bozicrHT** [llvm/llvm-project#219497]
**Description:**
Analyzer does not preserve the value written by `memset(c, 0, sizeof(*c))` when `c` is a pointer parameter. As a result, it considers `c->head != 0` feasible after the entire object was zeroed, producing a false `core.NullDereference` warning.

**Minimal reproducer:**
```c
#include <string.h>

void clang_analyzer_eval(int);

struct container { int *head; };

void test(struct container *c)
{
  int *tail = 0;

  memset(c, 0, sizeof(*c));

  clang_analyzer_eval(c->head == 0);
  // Expected: TRUE
  // Actual: TRUE and FALSE

  if (c->head != 0)
    *tail = 1; // false positive
}
```

Replacing `memset(c, 0, sizeof(*c))` with `c->head = 0` suppresses the warning. Compiler explorer reproducer can be found [here](https://compiler-explorer.com/z/KGvsGhbqc).

**Comment by steakhal** [comment]
Could you update the description with an inline and minimal example demonstrating the problem?

**Comment by bozicrHT** [comment]
Done. I simplified description and minimal reproducible example
